# Sheet

# Profitable App Profiles for the App Store and Google Play Markets

## DS PROJECT PHASES
1. **Business Understanding** - Ask relevant questions and define objectives for the problem.
2. **Data Mining** - Gather and scrape the data necessary for the project.
3. **Data Cleaning** - Fix the inconsistencies within the data and handle the duplicate and missing values.
4. **Data Exploration** - Form hypothesis about the defined problem by visually analyzing the data.
5. **Feature Engineering** - Select important features and construct more meaningful ones using the raw data.
6. **Data Science** - Perform data anlaysis to test the previously created hypothesis with respect to the problem definition.
7. **Data Visualization** - Communicate the findings with the key stakeholders using plots and interactive visualizations.

## 1. Business Understanding
### 📝 PROBLEM DEFINITION
As a business leader, I want to know what type of mobile apps can generate most revenue, so I can deliver higher developer time and effort to business value to the company stakeholders.

### ⛳️ GOAL
Identify mobile app profiles that are profitable for the Apple App Store and Google Play Store markets.

### 🔍 SCOPE
Our company only builds Android and iOS mobile apps for English speaking people that are free to download and install, and our main source of revenue consists of in-app ads. This means that our revenue for any given app is mostly influenced by the number of users that use our app.

## 2. Data Mining
> Since this is a sample project, we will work with an older data set. In a real-world scenario, we may have to spend time and resources to collect the latest data.

As of September 2018, there were approximately 2 million iOS apps available on the App Store, and 2.1 million Android apps on Google Play.

Collecting data for over four million apps requires a significant amount of time and money, so we'll try to analyze a sample of data instead. To avoid spending resources on collecting new data ourselves, we should first try to see whether we can find any relevant existing data at no cost. Fortunately, these are two data sets that seem suitable for our purpose:

- [A data set](https://www.kaggle.com/lava18/google-play-store-apps) containing data about approximately ten thousand Android apps from Google Play. You can download the data set directly from [this link](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv).
- [A data set](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) containing data about approximately seven thousand iOS apps from the App Store. You can download the data set directly from [this link](https://dq-content.s3.amazonaws.com/350/AppleStore.csv).

Let's start by opening the two data sets and then continue with the next phases.

In [1]:
from csv import reader

In [2]:
## Open Google Play Store Dataset
open_file = open('googleplaystore.csv')
read_file = reader(open_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

In [3]:
## Open Apple App Store Dataset
open_file = open('AppleStore.csv')
read_file = reader(open_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

Let's first take a quick look at the data and header to understand it. We will first start witht he Google Play Store data.

In [4]:
# let's create a function to print header with index number
# so we don't have to manually calculate it everytime when accessing the data
def header_indices(header):
    i = 0
    header_dict = {}
    while i < len(header):
        header_dict[header[i]] = i # with this format we can also use the col name to access the inxex number
        i += 1
    return header_dict

In [5]:
an_hi = header_indices(android_header)
an_hi

{'App': 0,
 'Category': 1,
 'Rating': 2,
 'Reviews': 3,
 'Size': 4,
 'Installs': 5,
 'Type': 6,
 'Price': 7,
 'Content Rating': 8,
 'Genres': 9,
 'Last Updated': 10,
 'Current Ver': 11,
 'Android Ver': 12}

In [6]:
android[:3]

[['Photo Editor & Candy Camera & Grid & ScrapBook',
  'ART_AND_DESIGN',
  '4.1',
  '159',
  '19M',
  '10,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'January 7, 2018',
  '1.0.0',
  '4.0.3 and up'],
 ['Coloring book moana',
  'ART_AND_DESIGN',
  '3.9',
  '967',
  '14M',
  '500,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design;Pretend Play',
  'January 15, 2018',
  '2.0.0',
  '4.0.3 and up'],
 ['U Launcher Lite – FREE Live Cool Themes, Hide Apps',
  'ART_AND_DESIGN',
  '4.7',
  '87510',
  '8.7M',
  '5,000,000+',
  'Free',
  '0',
  'Everyone',
  'Art & Design',
  'August 1, 2018',
  '1.2.4',
  '4.0.3 and up']]

In [7]:
# We're not going to use Pandas in this notebook but for the sake of better sneak peak at the data we will use it on the dataset
import pandas as pd

pd.DataFrame(android, columns=android_header)

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

      Rating Reviews                Size     Installs  Type Price  \
0        4.1     159                 19M      10,000+  Free     0   
1        3.9     967                 14M     500,000+  Free     0   
2        4.7   87510                8.7M   5,000,000+  Free     0   
3        4.5  215644                 25M  50,000,000+  Free     0   
4        4.3     967                2.8M     100,000+  Free     0   
...      ...     ...                 ...          ...   ...   ...   
10836    4.5      38                 53M       5,000+  Free     0   
10837    5.0       4                3.6M         100+  Free     0   
10838    NaN       3                9.5M       1,000+  Free     0   
10839    4.5     114  Varies with device       1,000+  Free     0   
10840    4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10841 rows x 13 columns]

## 3. Data Cleaning (DC)

1. Sometimes the dataset can have column shift due to incorrect columns of data for some rows. In google play store dataset one such row exist. It's helpful to write a helper function to do a quick check on such rows and eliminate them.
2. As shown in the statistics tab in the table above, we've 9660 unique apps out of 10841. That means we will need to eliminate duplicates.
3. As defined in the scope, our company only develops **free** apps. We can eliminate all paid apps from the data.
4. As defined in the scope, our company creates apps only for English-speaking people. We can filter out apps that have non-English characters in their app name.
5. Installs column will be important for our analysis as it represents the number of users (see problem definition), but it contains strings with "," and "+". Let's remove those two special characters and convert them to an `int` data type.

> **DC-1**

In [8]:
def clear_col_shift(data, length):
    dc_one = data.copy()
    for row in dc_one:
        if len(row) != length:
            dc_one.remove(row)
    return dc_one

In [9]:
android = clear_col_shift(android, len(android_header))

> **DC-2**
> 
> Before we consider removing duplicates, first we need to check why there are duplicates to identify which duplicate to keep and which to remove.

1. Let's observe some duplicates to spot any difference

In [10]:
unique_apps = []
duplicate_apps = []

for row in android:
    app = row[an_hi['App']]
    if app in unique_apps and app not in duplicate_apps:
        duplicate_apps.append(app)
    elif app not in unique_apps:
        unique_apps.append(app)

print(len(unique_apps))
print(len(duplicate_apps))

9659
798


In [11]:
print(android_header)
print('\n')
for row in android:
    app = row[an_hi['App']]
    if app == duplicate_apps[7] or app == duplicate_apps[-1]:
        print(row)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51510', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['AAFP', 'MEDICAL', '3.8', '63', '24M', '10,000+', 'Free', '0', 'Everyone', 'Medical', 'June 22, 2018', '2.3.1', '5.0 and up']
['AAFP', 'MEDICAL', '3.8', '63', '24M', '10,000+', 'Free', '0', 'Everyone', 'Medical', 'June 22, 2018', '2.3.1', '5.0 and up']


As we can see, the rows are almost similar except for the 'Reviews' column, which has sometimes slightly higher value. This is an observation based on small snapshot of the data, but it won't be a surprise to see a much higher value if the data of duplicate snapshots is different.

Keeping this in mind, along with the fact that more reviews is equal to more users, we will keep the duplicate rows with the highest reviews and delete all others.

In [12]:
# Approach 1 86.42% faster than approach 2 from Dataquest
import time
t1 = time.time()
# Create a list of duplicate with highest reviews count
duplicate_most = {}
for row in android:
    app = row[an_hi['App']]
    if app in duplicate_apps:
        if app in duplicate_most:
            if duplicate_most[app] < int(row[an_hi['Reviews']]):
                duplicate_most[app] = int(row[an_hi['Reviews']])
        else:
            duplicate_most[app] = int(row[an_hi['Reviews']])

dsp2 = []
added_apps = []
for row in android:
    app = row[an_hi['App']]
    reviews = int(row[an_hi['Reviews']])
    if app in duplicate_most:
        if reviews == duplicate_most[app] and app not in added_apps:
            added_apps.append(app)
            dsp2.append(row)
    else:
        dsp2.append(row)

t2 = time.time()

print("Time taken by Test 1: %.6f" %(t2 - t1))

Time taken by Test 1: 0.111777


In [13]:
# Aproach 2
# t1 = time.time()

# reviews_max = {}

# for app in android:
#     name = app[0]
#     n_reviews = float(app[3])
    
#     if name in reviews_max and reviews_max[name] < n_reviews:
#         reviews_max[name] = n_reviews
        
#     elif name not in reviews_max:
#         reviews_max[name] = n_reviews

# android_clean = []
# already_added = []

# for app in android:
#     name = app[0]
#     n_reviews = float(app[3])
    
#     if (reviews_max[name] == n_reviews) and (name not in already_added):
#         android_clean.append(app)
#         already_added.append(name) # make sure this is inside the if block

# t2 = time.time()

# print("Time taken by Test 2: %.6f" %(t2 - t1))

In [14]:
android = dsp2

> **DC-3**
>
> As defined in the scope, our company only develops free apps. We can eliminate all paid apps from the data.

In [15]:
free_apps = []

for row in android:
    if row[an_hi['Type']] == 'Free':
        free_apps.append(row)

android = free_apps

> **DC-4**
> 
> As defined in the scope, our company creates apps only for English-speaking people. We can filter out apps that have non-English characters in their app name.

In [16]:
# First let look at the data by sorting the app name field in ascending order, check both head and tail
pd.DataFrame(android, columns=android_header)

App             Category  \
0        Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1     U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
2                                 Sketch - Draw & Paint       ART_AND_DESIGN   
3                 Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
4                            Paper flowers instructions       ART_AND_DESIGN   
...                                                 ...                  ...   
8899                                   Sya9a Maroc - FR               FAMILY   
8900                   Fr. Mike Schmitz Audio Teachings               FAMILY   
8901                             Parkinson Exercices FR              MEDICAL   
8902                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
8903      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

     Rating Reviews                Size     Installs  Type Price  \
0       4.1     159                 19M      10,000+  Free     0   
1       4.7   87510                8.7M   5,000,000+  Free     0   
2       4.5  215644                 25M  50,000,000+  Free     0   
3       4.3     967                2.8M     100,000+  Free     0   
4       4.4     167                5.6M      50,000+  Free     0   
...     ...     ...                 ...          ...   ...   ...   
8899    4.5      38                 53M       5,000+  Free     0   
8900    5.0       4                3.6M         100+  Free     0   
8901    NaN       3                9.5M       1,000+  Free     0   
8902    4.5     114  Varies with device       1,000+  Free     0   
8903    4.5  398307                 19M  10,000,000+  Free     0   

     Content Rating                   Genres      Last Updated  \
0          Everyone             Art & Design   January 7, 2018   
1          Everyone             Art & Design    August 1, 2018   
2              Teen             Art & Design      June 8, 2018   
3          Everyone  Art & Design;Creativity     June 20, 2018   
4          Everyone             Art & Design    March 26, 2017   
...             ...                      ...               ...   
8899       Everyone                Education     July 25, 2017   
8900       Everyone                Education      July 6, 2018   
8901       Everyone                  Medical  January 20, 2017   
8902     Mature 17+        Books & Reference  January 19, 2015   
8903       Everyone                Lifestyle     July 25, 2018   

             Current Ver         Android Ver  
0                  1.0.0        4.0.3 and up  
1                  1.2.4        4.0.3 and up  
2     Varies with device          4.2 and up  
3                    1.1          4.4 and up  
4                    1.0          2.3 and up  
...                  ...                 ...  
8899                1.48          4.1 and up  
8900                 1.0          4.1 and up  
8901                 1.0          2.2 and up  
8902  Varies with device  Varies with device  
8903  Varies with device  Varies with device  

[8904 rows x 13 columns]

There are many apps with names that contain non-english characters. For example, other language characters, special characters, and emojis.

English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;), and other symbols (+, *, /)

Each character we use in a string has a corresponding number associated with it. For instance, the corresponding number for character 'a' is 97, character `'A'` is 65, and character `'爱'` is 29,233. We can get the corresponding number of each character using the `ord()` [built-in function](https://docs.python.org/3/library/functions.html#ord).

In [17]:
print(ord("a"))

97


The numbers corresponding to the characters we commonly use in an English text are all in the range `0 to 127`, according to the [ASCII](https://en.wikipedia.org/wiki/ASCII#Printable_characters) (American Standard Code for Information Interchange) system. Based on this number range, we can build a function that detects whether a character belongs to the set of common English characters or not. If the number is equal to or less than 127, then the character belongs to the set of common English characters.

If we're going to simply filter out name based on this approach, then we'll lose useful data since many English apps will be incorrectly labeled as non-English. To minimize the impact of data loss, we'll only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range. This means all English apps with up to three emojis or other special characters will still be labeled as English. Our filter function is still not perfect, but it should be fairly effective.

In [18]:
def is_english(name):
    non_ascii = 0
    for char in name:
        if ord(char) > 127:
            non_ascii += 1
    
    if non_ascii > 3:
        return False
    else:
        return True

In [19]:
english_apps = []

for row in android:
    app = row[an_hi['App']]
    if is_english(app):
        english_apps.append(row)

android = english_apps

> **DC-5**
> 
> **Installs** column will be important for our analysis as it represents the number of users (see problem definition), but it contains strings with `","` and `"+"`. Let's remove those two special characters and convert them to an int data type.

In [20]:
dc_5 = []

for row in android:
    installs = row[an_hi['Installs']]

    # convert installs to int
    installs = int(installs.replace('+', '').replace(',', ''))

    n_row = row.copy()
    n_row[an_hi['Installs']] = installs

    dc_5.append(n_row)

android = dc_5

## 4. Data Exploration

In [21]:
pd.DataFrame(android, columns=android_header)

App             Category  \
0        Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1     U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
2                                 Sketch - Draw & Paint       ART_AND_DESIGN   
3                 Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
4                            Paper flowers instructions       ART_AND_DESIGN   
...                                                 ...                  ...   
8858                                   Sya9a Maroc - FR               FAMILY   
8859                   Fr. Mike Schmitz Audio Teachings               FAMILY   
8860                             Parkinson Exercices FR              MEDICAL   
8861                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
8862      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

     Rating Reviews                Size  Installs  Type Price Content Rating  \
0       4.1     159                 19M     10000  Free     0       Everyone   
1       4.7   87510                8.7M   5000000  Free     0       Everyone   
2       4.5  215644                 25M  50000000  Free     0           Teen   
3       4.3     967                2.8M    100000  Free     0       Everyone   
4       4.4     167                5.6M     50000  Free     0       Everyone   
...     ...     ...                 ...       ...   ...   ...            ...   
8858    4.5      38                 53M      5000  Free     0       Everyone   
8859    5.0       4                3.6M       100  Free     0       Everyone   
8860    NaN       3                9.5M      1000  Free     0       Everyone   
8861    4.5     114  Varies with device      1000  Free     0     Mature 17+   
8862    4.5  398307                 19M  10000000  Free     0       Everyone   

                       Genres      Last Updated         Current Ver  \
0                Art & Design   January 7, 2018               1.0.0   
1                Art & Design    August 1, 2018               1.2.4   
2                Art & Design      June 8, 2018  Varies with device   
3     Art & Design;Creativity     June 20, 2018                 1.1   
4                Art & Design    March 26, 2017                 1.0   
...                       ...               ...                 ...   
8858                Education     July 25, 2017                1.48   
8859                Education      July 6, 2018                 1.0   
8860                  Medical  January 20, 2017                 1.0   
8861        Books & Reference  January 19, 2015  Varies with device   
8862                Lifestyle     July 25, 2018  Varies with device   

             Android Ver  
0           4.0.3 and up  
1           4.0.3 and up  
2             4.2 and up  
3             4.4 and up  
4             2.3 and up  
...                  ...  
8858          4.1 and up  
8859          4.1 and up  
8860          2.2 and up  
8861  Varies with device  
8862  Varies with device  

[8863 rows x 13 columns]